lib/core/src/chain/bitcoin/esplora.rs

```diff
@@ -0,0 +1,347 @@
+    }
+
+    // TODO Switch to batch search
+    async fn get_transactions(&self, txids: &[Txid]) -> Result<Vec<Transaction>> {
+        let client = self.get_client()?;
+        let mut result = vec![];
+        for txid in txids {
+            let tx = client
+                .get_tx(txid)
+                .await?
+                .context("Transaction not found")?;
+            result.push(tx);
+        }
+        Ok(result)
+    }
+
+    async fn get_script_history(&self, script: &Script) -> Result<Vec<History>> {
+        let client = self.get_client()?;
+        let history = client
+            .scripthash_txs(script, None)
+            .await?
+            .into_iter()
+            .map(|tx| History {
+                txid: tx.txid,
+                height: tx.status.block_height.map(|h| h as i32).unwrap_or(-1),
+            })
+            .collect();
+        Ok(history)
+    }
+
+    // TODO Switch to batch search
+    async fn get_scripts_history(&self, scripts: &[&Script]) -> Result<Vec<Vec<History>>> {
+        let mut result = vec![];
+        for script in scripts {
+            let history = self.get_script_history(script).await?;
+            result.push(history);
+        }
+        Ok(result)
+    }
+
+    async fn get_script_history_with_retry(
+        &self,
+        script: &Script,
+        retries: u64,
+    ) -> Result<Vec<History>> {
+        let script_hash = sha256::Hash::hash(script.as_bytes()).to_hex();
+        info!("Fetching script history for {}", script_hash);
+        let mut script_history = vec![];
+
+        let mut retry = 0;
+        while retry <= retries {
+            script_history = self.get_script_history(script).await?;
+            match script_history.is_empty() {
+                true => {
+                    retry += 1;
+                    info!(
+                        "Script history for {} got zero transactions, retrying in {} seconds...",
+                        script_hash, retry
+                    );
+                    tokio::time::sleep(Duration::from_secs(retry)).await;
+                }
+                false => break,
+            }
+        }
+        Ok(script_history)
+    }
+
+    async fn get_script_utxos(&self, script: &Script) -> Result<Vec<Utxo>> {
+        let client = self.get_client()?;
+        let utxos = client
+            .scripthash_outputs(script)
+            .await?
+            .into_iter()
+            .map(|out| {
+                Utxo::Bitcoin((
+                    OutPoint {
+                        txid: out.txid,
+                        vout: out.vout as u32,
+                    },
+                    TxOut {
+                        value: Amount::from_sat(out.value),
+                        script_pubkey: script.into(),
+                    },
+                ))
+            })
+            .collect();
+        Ok(utxos)
+    }
+
+    async fn get_scripts_utxos(&self, scripts: &[&Script]) -> Result<Vec<Vec<Utxo>>> {
```

Comment:
It would be nice to have some tests for the logic in this method, given its intricacies.